sycl/include/CL/sycl/detail/pi.hpp

```diff
@@ -165,24 +173,29 @@ namespace RT = cl::sycl::detail::pi;
 #define PI_TRACE(func) RT::Trace<decltype(func)>(func, #func)
 
 // This does the call, the trace and the check for no errors.
-#define PI_CALL(pi)                                                            \
-  RT::initialize(), RT::PiCall(#pi).check<cl::sycl::runtime_error>(            \
-                        RT::cast<detail::RT::PiResult>(pi))
+// Need to call initialize so that the ptrs are updated and then stored.
+#define PI_CALL(pi, ...)                                                       \
+  {                                                                            \
+    RT::initialize();                                                          \
+    RT::piCheckResult(PI_TRACE(pi)(__VA_ARGS__));                              \
```

Comment:
I don't think the check should be done here because there is no context to generate meaningful exception. Error checking and generation of appropriate exception should be done on the caller side.

Reply:
It is actually caller's decision to use PI_CALL or PI_CALL_RESULT or PI_CHECK_THROW depending on its context and needs. I think we should just document it better that these different ways exist for user's convenience.  When we switch to use a class to do the call, then the PI initialization would be moved in a constructor of that class.

Reply:
Ok, so we will need to convert all PI_CALL to PI_CALL_RESULT eventually.